clients/filesystem-fuse/src/default_raw_filesystem.rs

```diff
@@ -0,0 +1,398 @@
+                &opened_file.file_stat.path,
+            );
+        }
+
+        // put the file to the opened file manager
+        let opened_file = self.opened_file_manager.put(opened_file);
+        let opened_file = opened_file.lock().await;
+        Ok(opened_file.file_handle())
+    }
+
+    async fn create_dir(&self, parent_file_id: u64, name: &str) -> crate::filesystem::Result<u64> {
+        let parent_file_entry = self.get_file_entry(parent_file_id).await?;
+        let mut filestat = self.fs.create_dir(&parent_file_entry.path, name).await?;
+
+        filestat.set_file_id(parent_file_id, self.next_file_id());
+
+        // insert the new file to file entry manager
+        {
+            let mut file_manager = self.file_entry_manager.write().await;
+            file_manager.insert(filestat.parent_file_id, filestat.file_id, &filestat.path);
+        }
+        Ok(filestat.file_id)
+    }
+
+    async fn set_attr(&self, file_id: u64, file_stat: &FileStat) -> crate::filesystem::Result<()> {
+        let file_entry = self.get_file_entry(file_id).await?;
+        self.fs.set_attr(&file_entry.path, file_stat, true).await
+    }
+
+    async fn remove_file(&self, parent_file_id: u64, name: &str) -> crate::filesystem::Result<()> {
+        let parent_file_entry = self.get_file_entry(parent_file_id).await?;
+        self.fs.remove_file(&parent_file_entry.path, name).await?;
+
+        // remove the file from file entry manager
+        {
+            let mut file_manager = self.file_entry_manager.write().await;
+            file_manager.remove(&join_file_path(&parent_file_entry.path, name));
+        }
+        Ok(())
+    }
+
+    async fn remove_dir(&self, parent_file_id: u64, name: &str) -> crate::filesystem::Result<()> {
+        let parent_file_entry = self.get_file_entry(parent_file_id).await?;
+        self.fs.remove_dir(&parent_file_entry.path, name).await?;
+
+        // remove the dir from file entry manager
+        {
+            let mut file_manager = self.file_entry_manager.write().await;
+            file_manager.remove(&join_file_path(&parent_file_entry.path, name));
+        }
+        Ok(())
+    }
+
+    async fn close_file(&self, _file_id: u64, fh: u64) -> crate::filesystem::Result<()> {
+        let opened_file = self
+            .opened_file_manager
+            .remove(fh)
+            .ok_or(Errno::from(libc::EBADF))?;
+        let mut file = opened_file.lock().await;
+        file.close().await
+    }
+
+    async fn read(
+        &self,
+        _file_id: u64,
+        fh: u64,
+        offset: u64,
+        size: u32,
+    ) -> crate::filesystem::Result<Bytes> {
+        let file_stat: FileStat;
+        let data = {
+            let opened_file = self
+                .opened_file_manager
+                .get(fh)
+                .ok_or(Errno::from(libc::EBADF))?;
+            let mut opened_file = opened_file.lock().await;
+            file_stat = opened_file.file_stat.clone();
```

Comment:
no need to clone filestat?

Reply:
the `file_stat` is owned by the `openedFile`. we can't move it.

Reply:
can you use a reference?

Reply:
Cannot use a reference